### ISSUE TITLE
[Windows] Support Nightly Toolchains for main and release/5.10

### ISSUE THREAD

**Issue by shahmishal** [apple/swift#69484]

**Comment by shahmishal** [comment]
Currently blocked on https://github.com/apple/swift/issues/69483

**Comment by shahmishal** [comment]
cc: @grynspan @compnerd

**Comment by shahmishal** [comment]
Currently blocked on getting this PR merged https://github.com/apple/swift/pull/69555

**Comment by shahmishal** [comment]
Job created for main branch: https://ci-external.swift.org/job/swift-main-windows-toolchain/20/

**Comment by shahmishal** [comment]
Job failed to find `zlib` 
```
CMake Error: The source directory "C:/Users/swift-ci/jenkins/workspace/swift-main-windows-toolchain/zlib" does not exist.
```

Found the issue, and started new build: https://ci-external.swift.org/job/swift-main-windows-toolchain/21/console

**Comment by compnerd** [comment]
@shahmishal super exciting!

Just looking through a couple of things stand out:

> swift-ci@EC2AMAZ-JCVOGM7 C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain>C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\Tools\vsdevcmd.bat" -arch=amd64 -host_arch=amd64 
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.

With the switch to build.ps1, this shouldn't be needed (build.ps1 will properly setup the VS Dev environment to support building the SDK for all the architectures). Something that might help simplify the Jenkins configuration.

> swift-ci@EC2AMAZ-JCVOGM7 C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain>d /s /q C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\build 
'd' is not recognized as an internal or external command,
operable program or batch file.

Odd, I'm not sure where this is coming from.  We should investigate this later.

> swift-ci@EC2AMAZ-JCVOGM7 C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain>powershell.exe -ExecutionPolicy RemoteSigned -File C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\utils\build.ps1             -SourceCache C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain             -BinaryCache T:             -ImageRoot T:             -BuildType Release                           -Test swift,dispatch,foundation,xctest,             -Stage T:\artifacts 

This looks right ... and if this is failing, I doubt it's a configuration issue and more likely this is an issue in build.ps1. I'll need a little bit to dig into this a bit further to try to understand what is going on.

**Comment by compnerd** [comment]
There seem to be a few failures due to `%swift_driver_plain` being expanded to `main`. This would mean that `config.swift = inferSwiftBinary('swift')` returned `main`? I don't see that in the log, so I expect that it found it in the path.

**Comment by shahmishal** [comment]
If it helps, here is the build step: 

```
            "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\Common7\\Tools\\vsdevcmd.bat" -arch=amd64 -host_arch=amd64"

            rd /s /q %LocalAppData%\\clang\\ModuleCache


            set CMAKE_BUILD_TYPE=Release
            set PYTHON_HOME="%ProgramFiles(x86)%\\Microsoft Visual Studio\\Shared\\Python37_64"

            set SourceRoot=%WORKSPACE%
            set BuildRoot=%SourceRoot%\\build

            md %BuildRoot%
            subst T: /d
            subst T: %BuildRoot% || (exit /b)

            set BuildRoot=T:

            set PackageRoot=%BuildRoot%\\artifacts

            md %PackageRoot%

            :: Setup temporary directories
            md %BuildRoot%\\tmp
            set TEMP=%BuildRoot%\\tmp
            set TMP=%BuildRoot%\\tmp
            set TMPDIR=%BuildRoot%\\tmp

            set NINJA_STATUS=[%%f/%%t][%%p][%%es]

            set TestArg=-Test swift,dispatch,foundation,xctest,

            set "SkipPackagingArg= "

            rd /s /q %WORKSPACE%\\build

            powershell.exe -ExecutionPolicy RemoteSigned -File %WORKSPACE%\\swift\\utils\\build.ps1 ^
            -SourceCache %SourceRoot% ^
            -BinaryCache %BuildRoot% ^
            -ImageRoot %BuildRoot% ^
            -BuildType %CMAKE_BUILD_TYPE% ^
            %SkipPackagingArg% ^
            %TestArg% ^
            -Stage %PackageRoot%
```

**Comment by shahmishal** [comment]
>>swift-ci@EC2AMAZ-JCVOGM7 C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain>d /s /q C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\build
'd' is not recognized as an internal or external command,
operable program or batch file.

> Odd, I'm not sure where this is coming from. We should investigate this later.

I have a fix for this.

**Comment by compnerd** [comment]
It appears that a number of the failures are referencing `../lib/swift/windows` and not being able to find sources under there. I _think_ that this might be due to the split of the build tree and the source tree.

I wonder if there was something that we had done before that results in the build tree being linked or something?

**Comment by compnerd** [comment]
Hmm, I wonder if it is the `PYTHON_HOME` that is causing the issue ... we use python 3.9, but `PYTHON_HOME` is set to 3.7.  Would it be possible to test with `PYTHON_HOME` unset?

**Comment by compnerd** [comment]
Bad Run:
```
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:210: note: Using cmake: C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/bin/cmake.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-frontend: t:\1\bin\swift-frontend.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swiftc: t:\1\bin\swiftc.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using sil-opt: t:\1\bin\sil-opt.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using sil-func-extractor: t:\1\bin\sil-func-extractor.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using sil-llvm-gen: t:\1\bin\sil-llvm-gen.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using sil-nm: t:\1\bin\sil-nm.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using sil-passpipeline-dumper: t:\1\bin\sil-passpipeline-dumper.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using lldb-moduleimport-test: t:\1\bin\lldb-moduleimport-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-ide-test: t:\1\bin\swift-ide-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-dependency-tool: t:\1\bin\swift-dependency-tool.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-reflection-dump: t:\1\bin\swift-reflection-dump.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-remoteast-test: t:\1\bin\swift-remoteast-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-indent: t:\1\bin\swift-indent.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-symbolgraph-extract: t:\1\bin\swift-symbolgraph-extract.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-api-extract: t:\1\bin\swift-api-extract.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using clang: t:\1\bin\clang.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using clang++: t:\1\bin\clang++.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-link: t:\1\bin\llvm-link.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-llvm-opt: t:\1\bin\swift-llvm-opt.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-profdata: t:\1\bin\llvm-profdata.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-cov: t:\1\bin\llvm-cov.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-strings: t:\1\bin\llvm-strings.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using FileCheck: t:\1\bin\filecheck.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-dwarfdump: t:\1\bin\llvm-dwarfdump.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-readelf: t:\1\bin\llvm-readelf.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-dis: t:\1\bin\llvm-dis.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-nm: t:\1\bin\llvm-nm.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-readtapi: t:\1\bin\llvm-readtapi.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using llvm-size: t:\1\bin\llvm-size.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using sourcekitd-test: t:\1\bin\sourcekitd-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using complete-test: t:\1\bin\complete-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-api-digester: t:\1\bin\swift-api-digester.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-cache-tool: t:\1\bin\swift-cache-tool.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-refactor: t:\1\bin\swift-refactor.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-demangle-yamldump: t:\1\bin\swift-demangle-yamldump.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-demangle: t:\1\bin\swift-demangle.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:292: warning: couldn't find 'Benchmark_O' program, try setting BENCHMARK_O in your environment
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:292: warning: couldn't find 'Benchmark_Driver' program, try setting BENCHMARK_DRIVER in your environment
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using wasm-ld: t:\1\bin\wasm-ld.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-plugin-server: t:\1\bin\swift-plugin-server.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-parse-test: t:\1\bin\swift-parse-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:385: note: Using resource dir: ..\lib\swift
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:432: note: Compiling with -swift-version 4
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:528: note: Using Clang module cache: T:\1\swift-test-results\x86_64-unknown-windows-msvc\clang-module-cache
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:529: note: Using test results dir: T:/1/./swift-test-results/x86_64-unknown-windows-msvc
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:534: note: Using code completion cache: T:\1\swift-test-results\x86_64-unknown-windows-msvc\completion-cache
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:1496: note: Testing Windows x86_64-unknown-windows-msvc
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:2253: note: Using platform module dir: ..\lib\swift\windows
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:289: note: using swift-reflection-test-windows-x86_64: t:\1\bin\swift-reflection-test-windows-x86_64.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\test\lit.cfg:2861: note: Available features: CMAKE_GENERATOR=Ninja, CODEGENERATOR=AArch64, CODEGENERATOR=ARM, CODEGENERATOR=WebAssembly, CODEGENERATOR=X86, CPU=x86_64, MSVC_VER=17.0, OS=windows-msvc, PTRSIZE=64, SWIFT_VERSION=4, VENDOR=unknown, asserts, concurrency, concurrency_runtime, cplusplus_driver, differentiable_programming, distributed, executable_test, no_asan, no_lto, nonexecutable_test, objc_codegen, observation, optimized_stdlib, optimized_stdlib_x86_64, reflection, sourcekit, string_processing, swift_only_stable_abi, swift_stdlib_no_asserts, swift_swift_parser, swift_test_mode_optimize_none, swift_test_mode_optimize_none_x86_64, swift_tools_extra, thread_safe_runtime, threading_win32
```

Good Run:
```
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:210: note: Using cmake: C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/bin/cmake.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-frontend: t:\1\bin\swift-frontend.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift: t:\1\bin\swift.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swiftc: t:\1\bin\swiftc.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using sil-opt: t:\1\bin\sil-opt.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using sil-func-extractor: t:\1\bin\sil-func-extractor.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using sil-llvm-gen: t:\1\bin\sil-llvm-gen.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using sil-nm: t:\1\bin\sil-nm.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using sil-passpipeline-dumper: t:\1\bin\sil-passpipeline-dumper.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using lldb-moduleimport-test: t:\1\bin\lldb-moduleimport-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-ide-test: t:\1\bin\swift-ide-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-dependency-tool: t:\1\bin\swift-dependency-tool.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-reflection-dump: t:\1\bin\swift-reflection-dump.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-remoteast-test: t:\1\bin\swift-remoteast-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-indent: t:\1\bin\swift-indent.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-symbolgraph-extract: t:\1\bin\swift-symbolgraph-extract.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-api-extract: t:\1\bin\swift-api-extract.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using clang: t:\1\bin\clang.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using clang++: t:\1\bin\clang++.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-link: t:\1\bin\llvm-link.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-llvm-opt: t:\1\bin\swift-llvm-opt.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-profdata: t:\1\bin\llvm-profdata.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-cov: t:\1\bin\llvm-cov.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-strings: t:\1\bin\llvm-strings.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using FileCheck: t:\1\bin\filecheck.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-dwarfdump: t:\1\bin\llvm-dwarfdump.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-readelf: t:\1\bin\llvm-readelf.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-dis: t:\1\bin\llvm-dis.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-nm: t:\1\bin\llvm-nm.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-readtapi: t:\1\bin\llvm-readtapi.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using llvm-size: t:\1\bin\llvm-size.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using sourcekitd-test: t:\1\bin\sourcekitd-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using complete-test: t:\1\bin\complete-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-api-digester: t:\1\bin\swift-api-digester.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-cache-tool: t:\1\bin\swift-cache-tool.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-refactor: t:\1\bin\swift-refactor.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-demangle-yamldump: t:\1\bin\swift-demangle-yamldump.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-demangle: t:\1\bin\swift-demangle.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:292: warning: couldn't find 'Benchmark_O' program, try setting BENCHMARK_O in your environment
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:292: warning: couldn't find 'Benchmark_Driver' program, try setting BENCHMARK_DRIVER in your environment
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using wasm-ld: t:\1\bin\wasm-ld.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-plugin-server: t:\1\bin\swift-plugin-server.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-parse-test: t:\1\bin\swift-parse-test.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:385: note: Using resource dir: t:\1\lib\swift
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:432: note: Compiling with -swift-version 4
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:528: note: Using Clang module cache: T:\1\swift-test-results\x86_64-unknown-windows-msvc\clang-module-cache
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:529: note: Using test results dir: T:/1/./swift-test-results/x86_64-unknown-windows-msvc
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:534: note: Using code completion cache: T:\1\swift-test-results\x86_64-unknown-windows-msvc\completion-cache
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:1496: note: Testing Windows x86_64-unknown-windows-msvc
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:2253: note: Using platform module dir: t:\1\lib\swift\windows
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:289: note: using swift-reflection-test-windows-x86_64: t:\1\bin\swift-reflection-test-windows-x86_64.exe
lit.py: C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\test\lit.cfg:2861: note: Available features: CMAKE_GENERATOR=Ninja, CODEGENERATOR=AArch64, CODEGENERATOR=ARM, CODEGENERATOR=WebAssembly, CODEGENERATOR=X86, CPU=x86_64, MSVC_VER=17.0, OS=windows-msvc, PTRSIZE=64, SWIFT_VERSION=4, VENDOR=unknown, asserts, concurrency, concurrency_runtime, cplusplus_driver, differentiable_programming, distributed, executable_test, no_asan, no_lto, nonexecutable_test, objc_codegen, observation, optimized_stdlib, optimized_stdlib_x86_64, reflection, sourcekit, string_processing, swift_only_stable_abi, swift_stdlib_no_asserts, swift_swift_parser, swift_test_mode_optimize_none, swift_test_mode_optimize_none_x86_64, swift_tools_extra, thread_safe_runtime, threading_win32
```

This seems to be the diff:
```diff
- lit.py: swift\test\lit.cfg:289: note: using swift: t:\1\bin\swift.exe
- lit.py: swift\test\lit.cfg:385: note: Using resource dir: t:\1\lib\swift
+ lit.py: swift\test\lit.cfg:385: note: Using resource dir: ..\lib\swift
- lit.py: swift\test\lit.cfg:2253: note: Using platform module dir: t:\1\lib\swift\windows
+ lit.py: swift\test\lit.cfg:2253: note: Using platform module dir: ..\lib\swift\windows
```

This does match the majority of the failures. Now the question is, why is the resource dir and platform module dir path different?

**Comment by compnerd** [comment]
The missing

```
note: using swift: t:\1\bin\swift.exe
```

along with https://github.com/apple/swift/blob/main/test/lit.cfg#L273-L298

seems to indicate that `swift` was picked up from `Path`. Is the image identical to the PR builder? The default value is `..\lib\swift`.

**Comment by shahmishal** [comment]
@compnerd Let me know if I missed setting an env

**Comment by compnerd** [comment]
@shahmishal I don't see off hand any missing environment variables. The only ones that should be concerning are the ones that determine the flags (which repositories to test, whether to package or not, build type), but those seem to be setup.

I would say that removing the `PYTHON_HOME` setting might be good. One of the nice things about `build.ps1` is that we shouldn't need to set up a whole slew of environment variables, and if there is something missing we should try to detect it. If we could remove running the `VsDevCmd`, it would also simplify the run.

Just to double check my assumption - the image is identical to the PR testing one right?

**Comment by compnerd** [comment]
Hmm ... @tristanlabelle pointed out that I misread the code: [swift/test/lit.cfg at main · apple/swift](https://github.com/apple/swift/blob/main/test/lit.cfg#L275-L281)

It seems to imply that there is a `SWIFT` environment variable being set?

**Comment by shahmishal** [comment]
@compnerd  Can you try running this locally? 
```
 rd /s /q %LocalAppData%\\clang\\ModuleCache

 set CMAKE_BUILD_TYPE=Release
 set PYTHON_HOME="%ProgramFiles(x86)%\\Microsoft Visual Studio\\Shared\\Python37_64"

 set SourceRoot=%WORKSPACE%
 set BuildRoot=%SourceRoot%\\build

 md %BuildRoot%
 subst T: /d
 subst T: %BuildRoot% || (exit /b)

 set BuildRoot=T:

 set PackageRoot=%BuildRoot%\\artifacts

 md %PackageRoot%

 :: Setup temporary directories
 md %BuildRoot%\\tmp
 set TEMP=%BuildRoot%\\tmp
 set TMP=%BuildRoot%\\tmp
 set TMPDIR=%BuildRoot%\\tmp

 set NINJA_STATUS=[%%f/%%t][%%p][%%es]

 set TestArg=-Test swift,dispatch,foundation,xctest,

 set "SkipPackagingArg= "

 rd /s /q %WORKSPACE%\\build

 powershell.exe -ExecutionPolicy RemoteSigned -File %WORKSPACE%\\swift\\utils\\build.ps1 ^
 -SourceCache %SourceRoot% ^
 -BinaryCache %BuildRoot% ^
 -ImageRoot %BuildRoot% ^
 -BuildType %CMAKE_BUILD_TYPE% ^
 %SkipPackagingArg% ^
 %TestArg% ^
 -Stage %PackageRoot%
```

**Comment by compnerd** [comment]
@shahmishal everything now makes sense - there is an environment variable `swift` that is being set by Jenkins at [Parameters [Jenkins]](https://ci-external.swift.org/job/swift-main-windows-toolchain/21/parameters/#:~:text=n/test/lit.cfg%23L275-L281)%0A%0AIt%20seems%20to,variable%20being%20set?) that is set to `main`. That is then passed along to the test suite which then uses `main` as the `swift` command instead which results in the failures.

**Comment by shahmishal** [comment]
this is used to pick up the correct SHA, what is the best way to unset variable?

**Comment by compnerd** [comment]
Depends on the shell.  With `cmd` it would be `set swift=` (or whatever the variable name is). For PowerShell it would be `Delete-Item env:\swift`.

**Comment by shahmishal** [comment]
Going to try `set "swift= "`

**Comment by shahmishal** [comment]
New build started: https://ci-external.swift.org/job/swift-main-windows-toolchain/25/console

**Comment by compnerd** [comment]
@shahmishal if there is a trailing space, I think it might fail as it will set it to the string containing a space, there should be nothing after the `=`.

**Comment by compnerd** [comment]
> swift-ci@EC2AMAZ-JCVOGM7 C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain>set "swift= " 

The trailing space is setting the string to " " rather than deleting the environment variable.

**Comment by shahmishal** [comment]
main branch built: https://ci-external.swift.org/job/swift-main-windows-toolchain/26/

**Comment by compnerd** [comment]
Awesome!  Th 5.10 cherry-pick is at https://github.com/apple/swift/pull/69608

**Comment by shahmishal** [comment]
release/5.10 job created: https://ci-external.swift.org/job/swift-5.10-windows-toolchain/

**Comment by shahmishal** [comment]
```
swift-ci@EC2AMAZ-JCVOGM7 C:\Users\swift-ci\jenkins\workspace\swift-5.10-windows-toolchain>powershell.exe -ExecutionPolicy RemoteSigned -File C:\Users\swift-ci\jenkins\workspace\swift-5.10-windows-toolchain\swift\utils\build.ps1             -SourceCache C:\Users\swift-ci\jenkins\workspace\swift-5.10-windows-toolchain             -BinaryCache T:             -ImageRoot T:             -BuildType Release                           -Test swift,dispatch,foundation,xctest,             -Stage T:\artifacts 
Swift toolchain not found. Downloading from swift.org...
Installing Swift toolchain...
T:\toolchains\ : warning WIX8503: Failed to extract attached container. This most often happens when extracting a stripped bundle from the package cache, which is not supported.
Error: msiexec.exe exited with code 1619.
```

**Comment by shahmishal** [comment]
Removed `.` from the job name, let's see if it helps https://ci-external.swift.org/job/swift-5_10-windows-toolchain/5/console

**Comment by shahmishal** [comment]
@compnerd do you know whats going on?

**Comment by compnerd** [comment]
Hmm, not offhand.  Error 1619 is `ERROR_INSTALL_PACKAGE_OPEN_FAILED`.
> This installation package couldn't be opened. Verify that the package exists and is accessible, or contact the application vendor to verify that this is a valid Windows Installer package.

Can you verify if the path (`T:\toolchains\a0`) exists and is non-zero sized?  Either way, I think that we should hardcode the checksum for the expected binaries and ensure that they match.  I'll try to figure out how to do that in a separate change.

**Comment by compnerd** [comment]
I _think_ I might know what is going on.  There is a bug in the download URL, and so we were failing to download the installer.  This will need a little bit of help (we want to clean up the old 404 output). This should be caught in the future by the pinned SHA checks.

**Comment by shahmishal** [comment]
Do we know why this is only happening on `release/5.10` and not on `main`?

**Comment by compnerd** [comment]
I suspect that the URL was correct at one point and the artifact was cached.  Without being able to inspect the filesystem, it is difficult to understand what is going on.

**Comment by compnerd** [comment]
Put up https://github.com/apple/swift/pull/69663 to cherry-pick the changes to 5.10.

**Comment by shahmishal** [comment]
New build started with the cherry-pick merged into release/5.10 - https://ci-external.swift.org/job/swift-5_10-windows-toolchain/7/console

**Comment by compnerd** [comment]
Put up https://github.com/apple/swift/pull/69693 to cherry-pick some changes to help with the code signing.

**Comment by compnerd** [comment]
@shahmishal I don't think that there is in the current state.  If we had docker, we could easily build and test https://github.com/compnerd/swift-win32 as a smoke test (which is something that I do).

**Comment by shahmishal** [comment]
What is the status for hub.docker.com?

**Comment by compnerd** [comment]
Pinged the Docker folks again, all the feedback has been addressed.

**Comment by compnerd** [comment]
@shahmishal is that a blocker for getting the snapshots posted? I'm trying to understand where we are - we haven't had any snapshots for 5.10 yet, and I'm worried that we might get into a state where we cannot get the necessary changes merged and thus end up with a release on Windows that is not very usable.

**Comment by shahmishal** [comment]
No, it's not blocked on Windows integration testing or Docker image. 

The next step is for me to integrate the trigger from ci.swift.com to ci-external.swift.org. 

Plan on working on this integration in the next few days. Hopefully, it will be done before Nov 16th.

**Comment by shahmishal** [comment]
Hopefully, with this revert we will be able to get nightly toolchain unblocked. 
https://github.com/apple/swift/pull/69807

**Comment by compnerd** [comment]
The 5.10 toolchain is not usable due to missing files: https://github.com/apple/swift-installer-scripts/pull/258 should resolve that part.

**Comment by shahmishal** [comment]
New build being created for main branch for Windows nightly: https://ci-external.swift.org/job/swift-main-windows-toolchain/27/console

**Comment by compnerd** [comment]
@shahmishal hmm, that sounds like a download failure for the swift toolchain from swift.org?  That is very weird, I don't think that I've seen that since we fixed this in build.ps1 to do the validation of the downloaded image.

**Comment by shahmishal** [comment]
Rebuilding see if it helps https://ci-external.swift.org/job/swift-main-windows-toolchain/28/console

**Comment by shahmishal** [comment]
It failed again for same error: 
```
T:\toolchains\ : warning WIX8503: Failed to extract attached container. This most often happens when extracting a stripped bundle from the package cache, which is not supported.
Error: msiexec.exe exited with code 1619.
``

**Comment by compnerd** [comment]
@shahmishal we don't delete the installer if it existing, it might require some manual intervention to clean that up.

**Comment by shahmishal** [comment]
Where does the installer get installed? is it in T:\?

**Comment by compnerd** [comment]
Yeah, the installer itself is downloaded to `T:` but we extract the content into `T:\toolchains` IIRC

**Comment by shahmishal** [comment]
Can we delete this dir if it invalid and re-try extracting the toolchain?

**Comment by shahmishal** [comment]
@compnerd its still failing on another node, https://ci-external.swift.org/job/swift-main-windows-toolchain/29/console

**Comment by shahmishal** [comment]
I thought this would delete the virtual drive:
```
            subst T: /d
```

**Comment by compnerd** [comment]
Yes, that removes the "virtual drive", but does not remove the contents that were mapped there.

**Comment by shahmishal** [comment]
I also have `rd /s /q %WORKSPACE%\\build` in the job, so it should download new toolchain every time. 

```
            set BuildRoot=%SourceRoot%\\build

            md %BuildRoot%
            subst T: /d
            subst T: %BuildRoot% || (exit /b)
```

**Comment by compnerd** [comment]
Interesting - and the download doesn't say that the toolchain installer is invalid.  So, it could be that it is failing to pass the signature check :/ I don't think that I have a way to diagnose this without actually interacting with the host.

**Comment by shahmishal** [comment]
Can we add verbose mode to the signature check?

**Comment by shahmishal** [comment]
[msiexec_failure.log](https://github.com/apple/swift/files/13343448/log.log)
Attached log from 
```
msiexec.exe /lv T:\toolchains\log.log /qn /a T:\toolchains\a0 TARGETDIR=T:\toolchains\swift-5.9-RELEASE-windows10
```

**Comment by shahmishal** [comment]
```
bat '''
            "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\Common7\\Tools\\vsdevcmd.bat -arch=amd64 -host_arch=amd64"

            rd /s /q %LocalAppData%\\clang\\ModuleCache


            set CMAKE_BUILD_TYPE=Release
            set PYTHON_HOME="%ProgramFiles(x86)%\\Microsoft Visual Studio\\Shared\\Python37_64"

            set SourceRoot=%WORKSPACE%
            set BuildRoot=%SourceRoot%\\build

            md %BuildRoot%
            subst T: /d
            subst T: %BuildRoot% || (exit /b 1)

            set BuildRoot=T:

            set PackageRoot=%BuildRoot%\\artifacts

            md %PackageRoot%

            :: Setup temporary directories
            md %BuildRoot%\\tmp
            set TEMP=%BuildRoot%\\tmp
            set TMP=%BuildRoot%\\tmp
            set TMPDIR=%BuildRoot%\\tmp

            set NINJA_STATUS=[%%f/%%t][%%p][%%es]

            set TestArg=-Test swift,dispatch,foundation,xctest

            set "SkipPackagingArg= "
            set "swift="

            rd /s /q %WORKSPACE%\\build

            powershell.exe -ExecutionPolicy RemoteSigned -File %WORKSPACE%\\swift\\utils\\build.ps1 ^
            -SourceCache %SourceRoot% ^
            -BinaryCache %BuildRoot% ^
            -ImageRoot %BuildRoot% ^
            -BuildType %CMAKE_BUILD_TYPE% ^
            %SkipPackagingArg% ^
            %TestArg% ^
            -Stage %PackageRoot% || (exit /b 1)
            '''
```
Thats the current script, let me know if I missed something
Re-running it https://ci-external.swift.org/job/swift-main-windows-toolchain/31/console

**Comment by compnerd** [comment]
I wonder if the builders are running as different users.  `LOCAL SYSTEM` (service account) might encounter some issues - I wonder if we can quickly test with:

```
msiexec.exe /lvx! T:\toolchains\log.log /qn /a T:\toolchains\a0 TARGETDIR=T:\toolchains\swift-5.9-RELEASE-windows10 ALLUSERS=1
```

**Comment by compnerd** [comment]
https://github.com/apple/swift/pull/69852 adds the logging for future use as well.

**Comment by compnerd** [comment]
@shahmishal seems that the last re-run succeeded

**Comment by shahmishal** [comment]
@compnerd sent link in private for review for nightly main toolchain.

**Comment by compnerd** [comment]
@shahmishal - reviewed, seems ready to go to me!

**Comment by shahmishal** [comment]
`main` branch nightly Windows toolchain available on swift.org

**Comment by compnerd** [comment]
@shahmishal - it seems that the main branch nightlies do not auto-update?  There was a snapshot from 11/27, but the other platforms are on 12/7.

**Comment by compnerd** [comment]
@shahmishal any progress on the auto-updated nightlies? It seems that the Windows target is out of date by a month.

**Comment by compnerd** [comment]
ping again on this @shahmishal - the Windows snapshot on main is now ~2m out of date and we still do not have a 5.10 release snapshot.  This really risks the 5.10 release as we will need extra time to validate the Windows releases.

**Comment by shahmishal** [comment]
@compnerd I am looking into getting this configured, due to other tasks I was not able to get to this. Hopefully, I will be able to focus on this soon.

**Comment by shahmishal** [comment]
Started new builds
* 5.10 - https://ci-external.swift.org/job/swift-5.10-windows-toolchain/11/console 
* main - https://ci-external.swift.org/job/swift-main-windows-toolchain/33/console

**Comment by shahmishal** [comment]
main failed: 
```
swift-ci@EC2AMAZ-JCVOGM7 C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain>powershell.exe -ExecutionPolicy RemoteSigned -File C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\utils\build.ps1             -SourceCache C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain             -BinaryCache T:             -ImageRoot T:             -BuildType Release                           -Test swift,dispatch,foundation,xctest             -Stage T:\artifacts   || (exit /b 1 ) 
C:\Users\swift-ci\jenkins\workspace\swift-main-windows-toolchain\swift\utils\build.ps1 : A parameter cannot be found 
that matches parameter name 'BuildType'.
    + CategoryInfo          : InvalidArgument: (:) [build.ps1], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : NamedParameterNotFound,build.ps1
 
```
cc: @compnerd

**Comment by compnerd** [comment]
Ah, right, we dropped the `-BuildType` parameter a while ago to allow us to support sccache.  You should just be able to remove the `-BuildType Release` from the invocation (it will always generate release mode builds).

**Comment by shahmishal** [comment]
New build started: https://ci-external.swift.org/job/swift-main-windows-toolchain/34/console

**Comment by shahmishal** [comment]
> Ah, right, we dropped the `-BuildType` parameter a while ago to allow us to support sccache. You should just be able to remove the `-BuildType Release` from the invocation (it will always generate release mode builds).

Can we also cherry-pick this to release/5.10?

**Comment by compnerd** [comment]
Sure, let me look into that!

**Comment by shahmishal** [comment]
Nightly 5.10 released on swift.org - https://github.com/apple/swift-org-website/pull/488

**Comment by compnerd** [comment]
Trying to pull in the changes for `-BuildType` into 5.10 at https://github.com/apple/swift/pull/71117

**Comment by finagolfin** [comment]
[Swift.org now has builds from all the release branches](https://www.swift.org/install/windows/), so this can be closed?
